source/develop/release-management/features/virt/tpm-device.md

```diff
@@ -0,0 +1,154 @@
+---
+title: TPM devices
+category: feature
+authors: mzamazal
+feature_name: TPM devices
+feature_modules: engine,vdsm
+feature_status: In development
+---
+
+# TPM devices
+
+## Summary
+
+This feature allows adding emulated TPM (Trusted Platform Module) devices to VMs.
+
+## Owner
+
+*   Name: Milan Zamazal
+*   Email: mzamazal@redhat.com
+
+## Description
+
+TPM devices are useful to perform certain cryptographic operations (generating cryptographic keys, random numbers, hashes, etc.) or to store data that can be used to verify current software configuration securely.  QEMU and libvirt implement, among other, support for emulated TPM 2.0 devices, which is what oVirt uses to add TPM devices to VMs.
+
+Once an emulated TPM device is added to the VM, it can be used as a normal TPM 2.0 device in the guest OS.
+
+## Prerequisites
+
+There are no special prerequisites for this feature.  Software TPM emulation in QEMU is implemented with the help of [swtpm](https://github.com/stefanberger/swtpm) package, which is installed on the host automatically together with Vdsm.
+
+## Security
+
+TPM device data is managed by swtpm and stored in the host file system (until the VM stops running there) and in the Engine database (TODO: update according to the selected implementation).  Whoever has access to those locations has access to the TPM data.  This is necessary to provide the data to the currently running VM or on its next start.  libvirt provides an option to encrypt the data with a key, but the key still must be stored somewhere in Engine and in the domain XML, so it's more obfuscation rather than real security and it's currently not used by oVirt.
+
+Availability of data stored in TPM shouldn't be critical.  Its loss can cause operational problems such as a need to initialize and set up TPM data again, to audit the operating system after the TPM data is reset (lost) or perhaps even to reinstall the whole system but it shouldn't cause real data loss.  When using bare metal, one can't rely on TPM data availability either because if the hardware stops working, the data is no longer accessible.  If one uses for example a private key stored only in TPM to encrypt and decrypt file system encryption keys, then the risk of losing data stored in the file system without having proper backups is at least as big with hardware TPM as with virtual TPM.
+
+## Limitations
```

Comment:
Would we need to block operations like export/clone of running VMs with TCP devices (those operations take a snapshot without memory, only of the disks, and correlate it with the VM configuration at the time of the snapshot)?

Reply:
Did you mean "with TPM devices"? That's a good question, I'm afraid we can fall into trouble with all kinds of disk snapshots when the VM is running. A possible solution could be checking TPM data before starting the snapshot and before finalizing it and if it doesn't match then to let the snapshot fail.